### ISSUE TITLE
KCP rollout logs several condition patch conflicts

### ISSUE THREAD

**Issue by benmoss** [kubernetes-sigs/cluster-api#3395]
**What steps did you take and what happened:**
Triggered a rollout of my control plane by changing the KCP's spec.version property.

**What did you expect to happen:**
No conflicts should be happening

**Anything else you would like to add:**

This wasn't the first rollout I did on this cluster, so I think that may be why the `RollingUpdateInProgress` condition already existed.

```
controller.go:206] controllers/KubeadmControlPlane "msg"="Failed to patch KubeadmControlPlane" "error"="error patching conditions: The condition "Ready" was modified by a different process and this caused a merge/ChangeCondition conflict:
v1alpha3.Condition{
  	Type:               "Ready",
  	Status:             "False",
  	Severity:           "Warning",
- 	LastTransitionTime: v1.Time{Time: s"2020-07-24 15:06:24 +0000 UTC"},
+ 	LastTransitionTime: v1.Time{Time: s"2020-07-24 15:06:26 +0000 UTC"}
,
  	Reason:             "RollingUpdateInProgress",
- 	Message:            "Rolling 3 replicas with outdated spec (3 replicas up to date)",
+ 	Message:            "Rolling 3 replicas with outdated spec (0 replicas up to date)",
  }
controller.go:248] controller-runtime/controller "msg"="Reconciler error" "error"="error patching conditions: The condition "Ready" was modified by a different process and this caused a merge/ChangeCondition conflict:
v1alpha3.Condition{
  	Type:               "Ready",
  	Status:             "False",
  	Severity:           "Warning",
- 	LastTransitionTime: v1.Time{Time: s"2020-07-24 15:06:24 +0000 UTC"},
+ 	LastTransitionTime: v1.Time{Time: s"2020-07-24 15:06:26 +0000 UTC"},
  	Reason:
      "RollingUpdateInProgress",
- 	Message:            "Rolling 3 replicas with outdated spec (3 replicas up to date)",
+ 	Message:            "Rolling 3 replicas with outdated spec (0 replicas up to date)",
  }
controller.go:206] controllers/KubeadmControlPlane "msg"="Failed to patch KubeadmControlPlane" "error"="error patching conditions: The condition "MachinesReady" was modified by a different process and this caused a merge/ChangeCondition conflict:
v1alpha3.Condition{
  	Type:               "MachinesReady",
- 	Status:             "False",
+ 	Status:             "True",
- 	Severity:           "Warning",
+ 	Severity:           "",
  	LastTransitionTime: v1.Time{Time: s"2020-07-24 15:06:45 +0000 UTC"},
- 	Reason:             "InstanceNotReady@Machine/bmo-control-plane-8scgc",
+ 	Reason:             "",
- 	Message:            "3 of 4 completed",
+ 	Message:            "",
  }
controller.go:248] controller-runtime/controller "msg"="Reconciler error" "error"="error patching conditions: The condition "MachinesReady" was modified by a different process and this caused a merge/ChangeCondition conflict:
v1alpha3.Condition{
  	Type:               "MachinesReady",
- 	Status:             "False",
+ 	Status:             "True",
- 	Severity:           "Warning",
+ 	Severity:           "",
  	LastTransitionTime: v1.Time{Time: s"2020-07-24 15:06:45 +0000 UTC"},
- 	Reason:             "InstanceNotReady@Machine/bmo-control-plane-8scgc",
+ 	Reason:             "",
- 	Message:            "3 of 4 completed",
+ 	Message:            "",
  }
```

/kind bug
/area control-plane

**Comment by vincepri** [comment]
/milestone v0.3.x

I don't think we're updating these multiple times within the same reconciliation loop, it could be a stale cache issue that stays behind. Let's investigate a little more to see why this is happening.

**Comment by fabriziopandini** [comment]
If the problem is stale cache entries, a possible solution is to add `WithOwnedConditions` to https://github.com/kubernetes-sigs/cluster-api/blob/3c73ab1b8f793e9ca17d84283ed8adc6a502ad26/controlplane/kubeadm/controllers/controller.go#L205.
This should potentially be implemented for all the CAPI/CAPD controllers to avoid similar problems

**Comment by benmoss** [comment]
My initial instinct was that we should be fixing the root problem and not getting stale reads, but after digging through controller-runtime issues and docs it seems that this is just the [expected behavior](https://github.com/kubernetes-sigs/controller-runtime/blob/cea989b02ed02254d751ad30471c40bacd5888fc/doc.go#L85-L88), aka we don't get read-your-writes consistency.

My understanding of `WithOwnedConditions` is it tells the patch helper to clobber changes in the case of conflicts. I guess my fear here is that if we have a stale read, we might be writing conditions based on stale information, and so will we are clobbering **more** up-to-date information with **less** up-to-date information?

**Comment by ncdc** [comment]
If the patch helper updated the object being patched (`kcp` in this case) with the latest information returned by the apiserver after the `PATCH` call, would that help at all?

**Comment by vincepri** [comment]
I don't think we necessarily need to fix all of these in v0.3.x if we're planning to move to server side apply in v1alpha4

**Comment by vincepri** [comment]
/close

In favor of server-side apply in v1alpha4.
